engine/evd.mli

```diff
@@ -276,6 +276,15 @@ val add_quconstraints : evar_map -> Sorts.QUConstraints.t -> evar_map
 val undefined_map : evar_map -> undefined evar_info Evar.Map.t
 (** Access the undefined evar mapping directly. *)
 
+val undefined_evars : evar_map -> Evar.Set.t
+(** Access the set of undefined evars. *)
+
+val defined_map : evar_map -> defined evar_info Evar.Map.t
+(** Access the defined evar mapping directly. *)
+
+val defined_evars : evar_map -> Evar.Set.t
+(** Access the set of defined evars. *)
```

Comment:
Not sure we want to expose this stuff. What do you need it for?

Reply:
When coq-elpi returns a solution, it drops some evars that it created, so it needs to check whether the evar it wants to drop is reachable from the evars it has to keep. However, the evars from the input evar_map are not taken into account at the root of the search since we can not access them. This is just an accessor, so there should be no problem, right?

Reply:
Evd should be as small as possible, because the more stuff we have below the abstract interface, the more we can expose unsafe internals without realizing. My official position about the kind of evar surgery done by coq-elpi is that you shouldn't do that. There should be a GC interface exposed in the Evd API, and it shouldn't be the job of elpi to drop evars which migh break other invariants (reachability through evars is not the only way for an evar to be live).

Regardless of that general premise, I think that the set-returning functions shouldn't belong there, they're trivial composition with the domain function. So you should only expose defined_map, assuming this is even a good idea.

Reply:
I completely agree with this position, and having the API here would be great. But is there someone that could do it? I can open an issue if that helps.
I removed the set-returning functions, although I feel like they are the kind of things that are semantically important, and not having them makes the user code less readable.

Reply:
@ppedrot While I do agree that ideally one should expose a piece of API that is safe while doing the same job, I do not understand what is gained by minimizing an API that you deem unsafe: as soon as you expose one bit, you might as well expose its closure.
Until someone does the job of providing the safe API, I would say the best course of action is 1. tagging the API as unsafe in some way, and 2. exposing as many methods as needed so as to prevent users from doing even more unsafe stuff.